1차 리뷰 반영

102092 opened the pull request:
Resolved #82 

- 아직 까지 모든 리뷰 사항을 반영하진 않았지만, 일단 풀리퀘를 띄어놓고 보면서 상시적으로 보면서 진행하면 좋겠다는 생각에 풀리퀘를 열었습니다.

102092 committed:
Merge pull request #30 from Hydronium-ion/fix

[#29] Fix : application.yml 수정

102092 committed:
Merge pull request #39 from Hydronium-ion/dev

Author, ApiDocument

102092 committed:
Merge pull request #43 from Hydronium-ion/dev

CreatedTimeAt 변수명 변경

102092 committed:
Merge pull request #47 from Hydronium-ion/dev

CD script 수정

102092 committed:
Merge pull request #51 from Hydronium-ion/dev

썸네일 URL 랜덤으로

102092 committed:
Merge pull request #55 from Hydronium-ion/dev

Issue-033, issue-052

102092 committed:
Merge pull request #72 from Hydronium-ion/dev

8월 1주차

102092 committed:
[#79] Feat : method, origin 명시

- cors 설정에, 모든 것을 열지 않고 꼭 필요한 부분만 열도록 명시

102092 committed:
Merge pull request #80 from Hydronium-ion/fix

[#79] Feat : method, origin 명시

102092 committed:
[#79] Feat : local h2 add

102092 committed:
[#82] Chore : p6spy 의존성 삭제

- 서버 과부화 등을 유발할 수 있다는 조언을 받아서 삭제함

102092 committed:
[#82] Refactor : Post 정적 팩토리 메서드 변경

- Request DTO와 Account를 통해서 Post Entity를 생성하므로 이를 명확하게 나타내기 위해서 변경함

102092 committed:
[#82] Refactor : PostCreateRespons 생성자를 통해 하도록

- setter가 아닌 생성자를 통해 객체가 만들어지도록 수정함

102092 committed:
[#82] Refactor : 개행 정리

102092 committed:
[#82] Refactor : static 메소드 변경

- 여러 인자를 받는 만큼 of 로 명칭 변경

102092 committed:
[#82] Chore : RestAssured 의존성 추가

102092 committed:
[#82] Chore : 기본 profile 수정

102092 committed:
[#82] Refactor : 접근제어자 명시

- 코드 통일성을 위해서 접근 제어자를 명시함

102092 committed:
[#82] Refactor : main API

- get posts 일 떄는 likes 기준 내림차순으로 정렬
- posts/recent 는 createdAt 기준 내림차순 정렬로 변경
- enum을 이용해서, sort page reqeust를 만들 때, 하드코딩하던 부분을 제거

102092 committed:
[#82] Refactor : PostCreateRequest

- final 추가
- Builder 생성자에 추가

102092 committed:
[#82] Refactor : Account Entity 수정

- Builder 생성자에 추가

102092 committed:
[#82] Refactor : Post Entity 수정

- Builder 생성자에 추가

102092 committed:
Merge branch 'dev' into feature/issue-082

hanurii commented on build.gradle:
```diff
@@ -31,10 +31,12 @@ dependencies {
     testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc'
+    testImplementation 'org.springframework.restdocs:spring-restdocs-restassured'
     asciidoctor 'org.springframework.restdocs:spring-restdocs-asciidoctor'
-    testRuntimeOnly 'com.h2database:h2'
-    implementation 'com.github.gavlyukovskiy:p6spy-spring-boot-starter:1.5.6'
+    implementation 'com.h2database:h2'
```
1. p6spy 를 지운 신 이유가 있을까요? (그 때 말씀하시길 trace로 데이터를 확인 할 수 있어서 였나요?)
2. implementation 의 장점은 지난 번에 말씀 주셔서 숙지하고 있습니다.
H2 DB는 test 전용이니 testImplementation 은 어떠신지요!?
![주석 2020-08-11 141617](https://user-images.githubusercontent.com/38041572/89859982-3dd75900-dbdd-11ea-900d-906254af4cc9.png)
출처 : https://eblo.tistory.com/71

102092 replied:
### 1번
- sigrid의 조언, 로그를 남기는 건 편하지만 ec2에 부담이 될수도 있다는 조언 때문에 삭제하게 되었습니다. 가능하면 서버에 부담가는 부분은 줄이고 싶어서용..

### 2번
- 그렇네요 testImplementation이 좋겠어요
- 급하게 제 서버에 띄우다 보니까 바꿨던듯 같습니당.. 수정할게요!

hanurii commented on build.gradle:
```diff
@@ -31,10 +31,12 @@ dependencies {
-    testRuntimeOnly 'com.h2database:h2'
-    implementation 'com.github.gavlyukovskiy:p6spy-spring-boot-starter:1.5.6'
+    implementation 'com.h2database:h2'
+    implementation group: 'com.github.ulisesbocchio', name: 'jasypt-spring-boot-starter', version: '3.0.3'
```
jasypt는 삭제 해야 할 것 같습니다 (눈물..)

102092 replied:
😢  넹...

hanurii commented on src/main/java/com/codesquad/rare/config/Condition.java:
```diff
@@ -0,0 +1,19 @@
+
+  private String name;
+
+  Condition(String name) {
+    this.name = name;
+  }
+
+  public String getName() {
+    return name;
+  }
+}
```
- @Getter 가 있는데 getName 메서드의 존재이유는 무엇인가요?
- Condition 생성자를 @AllArgs... 로 변경하는 건 어떠신가요?

enum 클래스는 잘 사용하지 않아서 궁금증이 듭니다

102092 replied:
lombok의 존재를 모르고... getter 메서드를 명시해버렸네요.ㅠㅠ
수정하도록 하겠습니당..

hanurii commented on src/main/java/com/codesquad/rare/config/WebConfig.java:
```diff
@@ -32,9 +32,6 @@ public Jackson2ObjectMapperBuilderCustomizer jsonCustomizer() {
-        .allowedOrigins("*")
-        .allowedHeaders("*")
-        .allowCredentials(true);
+        .allowedOrigins("http://localhost:3000");
```
- CORS 에러 때문에 설정하신 걸로 알고 있는데 지워도 괜찮으신가요?
- 배포 때는 localhost:3000 은 변경하셔야 하시는 건가요?(궁금)

102092 replied:
- 너무 많은 부분을 개방하고 있는 듯해서, 최소 부분만 개방하기 위해서 수정했습니다.
- 추후 개발상황에 따라 추가적으로 개방하면 되지 않을까 생각합니다.
- 현재 master에도 똑같은 설정으로 배포되어 있는데.. 아마도 nginx 프론트 자원이 배포되거나 혹은 cloudFront에 배포되면 설정을 추가해야되지 않나 생각합니다.

hanurii commented on src/main/java/com/codesquad/rare/domain/account/Account.java:
```diff
@@ -27,4 +23,10 @@
 
   private String avatarUrl;
+
+  @Builder
+  private Account(String name, String avatarUrl) {
+    this.name = name;
+    this.avatarUrl = avatarUrl;
+  }
```
- @Builder를 없애시고 따로 밑에서 선언하신 이유는 무엇인가요?
- @AllArgsConstructor 삭제 이유는 무엇인가요?

102092 replied:
- sigirid의 조언에 따라, 생성자 위에 Builder를 명시해주기 위해서 수정하였습니다.
- 결국 해당 생성자에는 모든 인자들이 포함되니 AllArgsConstructor를 제거하였습니다.

hanurii commented on src/main/java/com/codesquad/rare/domain/post/Post.java:
```diff
@@ -69,7 +67,22 @@
+    this.createdAt = createdAt;
+  }
+
+  public static Post of(PostCreateRequest postCreateRequest, Account author) {
```
정적 팩토리 메서드 네이밍을 사용하셨군요 toEntity에 이제 익숙해졌는데... ㅋㅋㅋ 그래도 좋아요

hanurii commented on src/main/java/com/codesquad/rare/domain/post/PostController.java:
```diff
@@ -28,20 +31,18 @@
       @RequestParam(value = "page", required = false, defaultValue = DEFAULT_PAGE) int page,
       @RequestParam(value = "size", required = false, defaultValue = DEFAULT_SIZE) int size) {
-    return OK(postService.findAllAndOrderByCreatedAtDesc(page, size));
+    return OK(postService.findAll(LIKE.getName(), page, size));
```
Like.getName() 을 인수로 넘겨주는데 어떤 값이 들어가는지 잘 모르겠어요 설명 좀 부탁드립니다

hanurii replied:
밑에 코드를 보면서 이해했습니다. 굳굳

hanurii commented on src/main/java/com/codesquad/rare/domain/post/PostService.java:
```diff
@@ -23,8 +23,8 @@
 
-  public List<Post> findAllAndOrderByCreatedAtDesc(int page, int size) {
-    PageRequest pageRequest = PageRequest.of(page, size, Sort.by("createdAt").descending());
+  public List<Post> findAll(String condition, int page, int size) {
+    PageRequest pageRequest = PageRequest.of(page, size, Sort.by(condition).descending());
```
개인적인 생각입니다

findAll 이 인수로 String condition을 받고 그게 PageRequest 안에 Sort.by 안에 들어가는데요.
그냥 Sort.by 안에 Condition.LIKE (이렇게 표기하는게 맞나 모르겠네요) 표기하는 게 더 좋지 않을까요?

그 때 단이 리뷰 해줄 때 findAllByLikes() 보다는 findAll(Like like) 로 하는 게 좋을 것 같다고 한 걸 본 것 같습니다.
그 의견이 한 줄을 봤을 때 명시적이고 가독성이 좋으면 더 괜찮을 것 같다라고 생각을 했는데용
그래서  27라인을 그냥 보기만 한다면 이게 likes 정렬인지 createdAt 정렬인지 잘 모를 것 같아서 피드백 드립니다!!

102092 replied:
- 이 부분은 현재 controller에서 처리하고 있습니다.
```java
  @GetMapping
  public ApiResult<List<Post>> findAllByLikes(
      @RequestParam(value = "page", required = false, defaultValue = DEFAULT_PAGE) int page,
      @RequestParam(value = "size", required = false, defaultValue = DEFAULT_SIZE) int size) {
    return OK(postService.findAll(LIKE.getName(), page, size));
  }

  @GetMapping("/recent")
  public ApiResult<List<Post>> findAllByCreatedAt(
      @RequestParam(value = "page", required = false, defaultValue = DEFAULT_PAGE) int page,
      @RequestParam(value = "size", required = false, defaultValue = DEFAULT_SIZE) int size) {
    return OK(postService.findAll(CREATED.getName(), page, size));
  }
```
- 우선 velog의 메인 페이지가 기본적으로 `/` or `/recent ` 이런식으로 되어 있더라구요..
- 이를 따라가는 게 맞다고 생각되어 위와 같이 구성하였습니다.
- service 단에서 findAll이 위와 같이 한 메서드로 표현되고, controller에서 condition이 명시되어 있는 형태입니다.
- 어떤 것이 좋을 지는 좀 더 같이 얘기해보면 좋을 것 같아요
- 어떤 방식이 좋을ㄹ까요?

hanurii commented on src/main/java/com/codesquad/rare/domain/post/request/PostCreateRequest.java:
```diff
@@ -2,33 +2,39 @@
 
   @NotBlank(message = "부제목을 반드시 입력해주세요")
-  private String subTitle;
+  private final String subTitle;
 
-  private String content;
+  private final String content;
 
-  private String thumbnail;
+  private final String thumbnail;
 
   @NotNull(message = "작성자의 ID는 반드시 입력해주세요")
-  private Long authorId;
+  private final Long authorId;
 
-  private String tags;
+  private final String tags;
 
```
final로 선언하신 이유가 있을까여??

102092 replied:
- DTO에서는 멤버변수들이 final로 선언되는게, 추후 의도치 않는 데이터 변경을 막기 위해서 필요하지 않을까 생각했습니다.

hanurii commented on src/main/java/com/codesquad/rare/domain/post/response/PostCreateResponse.java:
```diff
@@ -1,16 +1,18 @@
 
   private Long postId;
 
+  public PostCreateResponse(Long postId) {
+    this.postId = postId;
+  }
```
@AllArgs... 는 어떠신가요

102092 replied:
좋습니다!

hanurii commented on src/main/resources/application.yml:
```diff
@@ -75,3 +75,41 @@ management:
     health:
       show-details: always
+
+---
+# spring, datasource
+spring:
+  profiles: fix
+  datasource:
+    driver-class-name: org.h2.Driver
+    url: jdbc:h2:mem:testdb;DB_CLOSE_ON_EXIT=FALSE
+    username: sa
+    password:
+  jpa:
+    database-platform: org.hibernate.dialect.H2Dialect
+    open-in-view: true
+    hibernate:
+      ddl-auto: create-drop
+    properties:
+      hibernate:
+        format_sql: true
+
+# logging
+logging:
+  level:
+    web: debug
+    com.codesquad.rare: debug
+    org.hibernate.SQL: debug
+    org.hibernate.type.descriptor.sql: trace
+
+# actuator
+management:
+  endpoints:
+    web:
+      exposure:
+        exclude: env,beans
+        include: "*"
+      base-path: '/system/actuator'
+  endpoint:
+    health:
+      show-details: always
```
main/application.yml 에 이걸 추가하실려고 build.gradle에서 implementation으로 구현하신건가요?? 👀

102092 replied:
이 부분은, 제가 급하게 서버 작동을 위해 추가한 부분입니다.
남아있었네요.. ㅠㅠ 삭제해서 다시 올리겠습니다.

hanurii commented on src/test/java/com/codesquad/rare/domain/post/PostControllerTest.java:
```diff
@@ -56,25 +56,25 @@
 
   @Autowired
-  MockMvc mockMvc;
+  private MockMvc mockMvc;
 
   @MockBean
-  PostController postController;
+  private PostController postController;
 
-  Random random = new Random();
+  private Random random = new Random();
 
-  Account won = Account.builder()
+  private Account won = Account.builder()
```
원래 안쓰신다고 하셨는데 접근제어자를 붙이신 이유가 있을까요??

102092 replied:
단의 리뷰를 통해 변경하게 되었습니다.
여전히 저는 안 써도 상관 없다고 생각하긴 하는데, 코드 통일성을 위해서 그리고 혹시 모를 사이드 이펙트를 줄이기 위해서
명시해주는게 안전하지 않을까 생각하게 되었습니다.

hanurii commented:
정말 고생하셨습니다 바쁘시겠지만 재리뷰 부탁드립니다

hanurii commented:
그리고 몇 가지 의견을 물어보고 싶은 게 있습니다.

1. NotFoundException 말고 BadRequestException 추가 의견
써니가 준 의견인데요 NotFoundException 말고 잘못된 요청이 왔을 때 뱉은 BadRequestException 도 있으면 좋겠다고 말을 했는데 사실 저는 익셉션 잘 활용을 못해서 뭐라고 답변을 못줬습니다. 한님 생각은 어떠신가요?

2. 저희가 테스트 메서드명을 모두 스네이크 케이스로 작성하고 있는데요 카멜케이스와 스네이크 케이스 중 어떤 게 가독성이 좋고 테스트 메서드명에 적합한지 궁금증이 생겨 좀 찾아봤습니다.
우선 정답은 둘다 사용하라였고 예를 들어 이런 형태였습니다.
```
- 가독성이 좋아짐 : createUser_WithNonExistingEmail_ShouldThrowArgumentException
- 가독성이 떨어짐 : createUserWithNonExistingEmailShouldThrowArgumentException
```
https://softwareengineering.stackexchange.com/questions/237561/naming-test-methods-in-java

102092 committed:
[#82] Chore : jaspty 의존성 삭제, h2 implementation 변경

102092 committed:
[#82] Chore : fix profile 삭제

102092 committed:
Merge remote-tracking branch 'remotes/origin/dev' into feature/issue-082

# Conflicts:
#	build.gradle

102092 committed:
[#82] Refactor : lombok annotation 사용을 통한 코드 최적화

- 롬복 사용을 통해, 명시하지 않아도 되는 코드를 삭제

102092 committed:
[#82] Refactor : Builder 수정

- CI 문제 해결하기 위해서, Entity에 명시된 Builder의 매개변수들을 추가함
-

102092 committed:
[#82] Refactor : enum Condition 패키지 수정

-  사용하는 패키지에 위치하도록 수정

102092 commented:
1. NotFoundException 말고 BadRequestException 추가 의견
- 예외를 좀 세분화 하자는 의견으로 이해됩니다.
- 좋은 의견이라 생각합니다. 예외 처리를 잘 할 수록 에러를 잡기 좋아지고, 프로젝트의 완성도가 높아진다고 생각해요.
- 다만, 현재 저희 단계에서는 굳이 나누지 않아도 된다고 생각해요.
- 추후 진행하면서, 예외를 세분화 해보면 좋을듯 합니다.

2. 테스트명
- 저는 스네이크 더 좋네요. `_` 를 더하면, 저에겐 가독성이 좋아지는 건 확실하네요
- 하밀은 어떤가요?

102092 committed:
Merge remote-tracking branch 'origin/feature/issue-082' into feature/issue-082

# Conflicts:
#	build.gradle

102092 committed:
[#82] Refactor : PostCreateResponse lombok 사용

- 명시된 생성자를 제거하고 롬복에서 지원하는 어노테이션 사용

102092 committed:
[#82] Refactor : PostService 매개 변수 final Option 추가

102092 committed:
[#82] Refactor : 접근제어 default를 수정

- 통일성 주기 위해서

102092 committed:
[#82] Refactor : Post 생성자 매개 변수를 final로 선언

102092 committed:
[#82] Refactor : Post DTO, 생성자 매개변수 수정

hanurii committed:
[#61] Feat : PostMainResponse 추가에 따른 코드 변경

- 메인 블로그 페이지에서 필요한 포스트 목록 필드와 개인 블로그 페이지에서 필요한 포스트 목록 필드에 차이가 있음
- 차이를 주고자 메인 블로그 페이지에 필요한 DTO를 생성함

hanurii committed:
[#61] Feat : 서비스 로직에서 Post 엔티티를 PostMainDTO로 변환

- DTO 에서 빌더 추가
- Service 로직에서 포스트 목록 출력하는 메서드의 반환 값과 로직 변경

hanurii committed:
[#61] Feat : 서비스 로직 변경에 따른 수정

- 서비스 로직에서 메서드 하나로 좋아요순과 최신순 컨트롤러 2개와 연결되어 있기 때문에
- 서비스 로직 변경에 따른 컨트롤러 로직 변경

hanurii committed:
[#61] Feat : 포스트 상세페이지 DTO 생성 후 서비스 로직 변경

hanurii committed:
[#102] Feat : 개인 블로그 진입 시 해당 유저의 포스팅한 포스트 출력

hanurii committed:
Refactor : PostCreateResponse 를 PostIdResponse 로 변경하여 다른 곳에서도 사용 가능하게 리팩토링

hanurii committed:
[#103] Feat : 포스트 수정 기능 구현

- 포스트를 수정 할 수 있습니다

hanurii committed:
[#68] Feat : 이미지 업로드 기능 구현

- 이미지 업로드 시 s3 로 업로드 됩니다
- isPostImage(boolean) 값이 true 이면 posts 디렉토리 밑으로, false면 users 디렉토리 밑에 이미지가 저장됩니다

hanurii committed:
[#68] Feat : 이미지 파일을 보내지 않았을 때 익셉션을 발생시킴

- 요청 시 이미지 파일이 비어있을 경우 익셉션 발생

hanurii committed:
[#68] Feat : 이미지 업로드 할 때 용량 제한

- 용량에 제한을 둠으로써 s3 를 과도하게 사용하는 것을 방지한다
- 용량은 10MB로 설정해뒀지만 추후 너무 S3 요금이 방대하게 나온다면 줄일 생각도 있다

hanurii committed:
[#68] Feat : 이미지 API 반환 값으로 S3 엔드포인트를 전달

- 이 엔드포인트 값을 이용해 글 작성 시 이미지 마크다운에 사용할 예정

hanurii committed:
[#68] Feat : 기본 암호화 AES-256 활성화

- S3 에 업로드 시 AES-256 암호화 세팅되지 않은 객체는 업로드 될 수 없다

hanurii committed:
[#68] Feat : 이미지 파일 확장자가 아닌 파일 확장자는 올릴 수 없는 로직 추가

- 이미지 관련 파일 확장자만 업로드 할 수 있도록 로직 추가
- 추후 기회가 된다면 동영상 파일 확장자도 추가 할 수도 있음

hanurii committed:
[#68] Feat : s3 업로드 과정 중에 잘못되면 보여줄 익셉션 구현

- 파일 확장자 익셉션 완성
- 파일 크기가 max 이상이면 발생하는 익셉션 완성
- s3 업로드 과정 중 업로드가 되지 않았을 때 발생하는 익셉션 구현

hanurii committed:
[#68] Feat : 중복 데이터 방지를 위해 fileName 포맷을 변경

- 중복 데이터 삽입 시 overwrite 될 수 있기 때문에 fileName 포맷을 변경
- overwrite 되지 않게 하는 설정이 s3에 있다고 하는데 우선 코드로 구현하는 방향을 선택

hanurii committed:
Feat : 컨트롤러 로직 변경으로 인한 테스트 코드 변경

- 컨트롤러의 리턴값을 엔티티가 아닌 DTO로 설정
- 이 후 변경된 사항에 대해 테스트 코드도 변경
